### PR TITLE
Notion pipeline

### PR DESCRIPTION
**Pull request by burnash**
Implementing a verified pipeline 

# Relevant issue

issue #136

### CODE REVIEW

File: pipelines/notion/client.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        return response.json()
+
+    def search(
+        self,
+        query: Optional[str] = None,
+        filter_criteria: Optional[Dict[str, Any]] = None,
+        sort: Optional[Dict[str, Any]] = None,
+        start_cursor: Optional[str] = None,
+        page_size: Optional[int] = None,
+    ) -> Dict[str, Any]:
```
Comment (rudolfix) [inline pipelines/notion/client.py:114]: this is Generator so correct type would be `Iterator[...]`.

if you want your pipeline to be type-checked, place a file called `py.typed` in the `./notion` folder. this is how our Makefile selects pipelines for type checking

File: pipelines/notion/client.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                'page_size': page_size,
+            }
+
+            filtered_payload = self._filter_out_none_values(payload)
+
+            response = self.send_payload(
+                'search', '', payload=filtered_payload
+            )
+
+            for result in response.get('results', []):
```
Comment (rudolfix) [inline pipelines/notion/client.py:151]: it is way faster to yield lists, not item by item. `dlt` will handle both lists and items same way, transparently for the user. list goes through pipeline as single item. if you use transformers, it is really way more efficient

File: pipelines/notion/database.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        while has_more:
+            response = self.notion_client.send_payload(
+                'databases',
+                self.database_id,
+                subresource='query',
+                query_params=filter_properties,
+                payload=payload,
+            )
+
+            for result in response.get('results', []):
```
Comment (rudolfix) [inline pipelines/notion/database.py:75]: same thing

**Review by rudolfix — CHANGES_REQUESTED** [review]
Thanks for the test, docs and docstrings, this is really high quality!

regarding the mock tests. I see some use for them for example to reproduce some edge cases from  customers that are hard to add to our test account. we can discuss how to use them

still I see a lot of value in testing on real account. ie. authorization, api changes, adding new test cases is IMO easier.

as we have many test in mocks, you can add just one test where we load a few tables from notion. please try many data types for tables (in Notion) so we can see how they are mapped to python values

File: pipelines/notion/client.py (hunk 2 above)
Comment (burnash) [reply pipelines/notion/client.py:151]: I use `search()` in `notion_databases()` to filter database ids myseld here:

```python
if database_ids is None:
    search_results = notion_client.search(
        filter_criteria={'value': 'database', 'property': 'object'}
    )
    database_ids = [
        {'id': result['id'], 'use_name': result['title'][0]['plain_text']}
        for result in search_results
    ]
```

If I yield a list, the usability of `search()` could suffer, since the calling code would need to handle chaining. For this particular case I'm not sure if the performance gains would be that visible. Happy to adjust if I'm wrong.

File: pipelines/notion/__init__.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import dlt
+from dlt.extract.source import DltResource
+
+from .client import NotionClient
+from .database import NotionDatabase
+
+
+@dlt.source
+def notion_databases(
+    database_ids: Optional[List[Dict[str, str]]] = None,
```
[the remark this answers is not in the document]
Comment (burnash) [reply pipelines/notion/__init__.py:14]: > if you set it to dlt.config.value the argument is required and None value is not allowed

What's the optimal way to let the value be None? The way I put it, when the `database_ids` is missing (and defaults to None), `notion_databases` queries all available databases.

[the remark this answers is not in the document]
Comment (rudolfix) [reply pipelines/notion/__init__.py:14]: ahhh then it is good. nothing to change there. what I mean is that you can still use `config.toml` to pass `database_ids`

File: pipelines/notion/client.py (hunk 2 above)
Comment (rudolfix) [reply pipelines/notion/client.py:151]: yes you are right (I feel like GPT-4 now ;p) you are using search not to query data, just the database ids... then all is good and we should keep yielding single elements

**Review by rudolfix — APPROVED** [review]
